sqlglot/optimizer/qualify_columns.py

```diff
@@ -623,15 +631,29 @@ def _add_except_columns(
         except_columns[id(table)] = columns
 
 
+def _add_rename_columns(
+    expression: exp.Expression, tables, rename_columns: t.Dict[int, t.Dict[str, str]]
+) -> None:
+    rename = expression.args.get("rename")
+
+    if not rename:
+        return
+
+    columns = {e.this.name: e.alias for e in rename}
+
+    for table in tables:
+        rename_columns[id(table)] = columns
+
+
 def _add_replace_columns(
-    expression: exp.Expression, tables, replace_columns: t.Dict[int, t.Dict[str, str]]
+    expression: exp.Expression, tables, replace_columns: t.Dict[int, t.Dict[str, exp.Alias]]
 ) -> None:
     replace = expression.args.get("replace")
 
     if not replace:
         return
 
-    columns = {e.this.name: e.alias for e in replace}
+    columns = {e.alias: e for e in replace}
```

Comment:
Do we actually want an `Alias` here as the value? Should we just keep the expression and let `qualify_outputs` take care of the aliasing?

Reply:
Hmm, not sure if that's feasible or better. What I had in mind was that `_expand_stars()` wants to produce a qualified projection list, in which case if a column `name` is in `replaced_columns` then it was defined as `REPLACE(<expr> AS name, ...)` so you can append that `exp.Alias` directly in`new_selections` instead of re-building it later.

Also, a potential side effect of keeping just the `e.this` expression is that `qualify_outputs` depends on the `e.output_name`, so I think such scenarios can happen:

```
query: WITH x AS (SELECT 1 AS a) SELECT * REPLACE(2 AS a) FROM x

after expand_stars():  WITH x AS (SELECT 1 AS a) SELECT 2 FROM x AS x

after qualify_outputs():  WITH x AS (SELECT 1 AS a) SELECT 2 AS "2" FROM x AS x
```

What do you think?

Reply:
Ok sounds good, the `"2"` example makes sense.

Reply:
Let's add a test for this btw